File: Trie/rcc.py
class trienode:
    
    def __init__(self,size):
        self.eow = False
        self.s = size
        self.child = [None]*size
        

    
def printvec(l):
    for i in l:
        print(i,end="")


def insert(T,m,string):
    idx = 0
    tmp = T
    
    for i in range(len(string)):
        idx = m.index(string[i]) - 0
    
        if tmp.child[idx] == None:
            tmp.child[idx] = trienode(len(m))
        tmp = tmp.child[idx] 

    tmp.eow = True

def printTr(T,m,stk):
    if (T==None or T.eow):
        printvec(stk)
        print()

        if T==None:
            for i in range(T.s):
                if (i):
                    print(m[i])
            print(end = "")

        return

    for i in range(T.s):
        if (T.child[i]):
            stk.append(m[i])
            printTr(T.child[i],m,stk)
            stk.pop()



if __name__=="__main__":
    o = False
    vec = []
    m = set()

    while True:
        s = str(input())
        if (s[-1]=='#'):
            o = True
            s = s[:-1]
        vec.append(s)
        for i in s:
            m.add(i)
        if o:
            break

    keys = list()
    m = sorted(list(m))
   
    T = trienode(len(m))
    stk = list()

    for i in vec:
        insert(T,m,i)
        
    print("TRIE : ")
    printTr(T,m,stk)

'''
abd5c2
ac8b9
cdd294a1
da5b4c63
bd367#
'''

File: Trie/rcc.cpp
#include <iostream>
#include <map>
#include <algorithm>
#include <vector>
#include <string>

using namespace std;

class trienode
{
public:
    bool EOW;
    int s;
    trienode **child;
    trienode(int size)
    {
        s = size;
        child = new trienode *[size];
	    EOW = false;
	    for (int i=0;i<size;++i) this->child[i] = NULL;
    }
};

void printvec(vector<char>v)
{
    for (int i=0;i<v.size();++i) cout<<v[i];
}

void insert(trienode* &T ,vector<char>keys,string str)
{
    int index;trienode* tmp = T;
    vector<char>::iterator it;
    int i;
    for (i = 0;i<str.length();++i)
    {
        index = find(keys.begin(),keys.end(),str[i]) - keys.begin();
        if (!tmp->child[index])
            tmp->child[index] = new trienode(keys.size());

        tmp = tmp->child[index];
    }
    //index = find(keys.begin(),keys.end(),str[i]) - keys.begin();
    //tmp[index] = true;
    tmp->EOW = true;
}

bool search(trienode*T,vector<char>keys,string str)
{
    int index,i=0;
    trienode* tmp = T;

    vector<char>::iterator it;
    for (i=0;i<str.length()-1;++i)
    {
        index = find(keys.begin(),keys.end(),str[i]) - keys.begin();
        if (!tmp->child[index]) 
            return false;

        tmp = tmp->child[index];
    }
    index = find(keys.begin(),keys.end(),str[i]) - keys.begin();

    return (tmp->EOW && tmp!=NULL);
}

void print(trienode*T,vector<char>keys,vector<char>stk)
{
    if (T==NULL || T->EOW)
    {
        printvec(stk);
        if (T!=NULL)
        {
            for (int i=0;i<T->s;++i) if(T->child[i]) cout<<keys[i];
        } 
        cout<<" ";
        return;
    }
    

    for (int i=0;i<T->s;++i)
    {
        if (T->child[i])
        {
            stk.push_back(keys[i]);
            print(T->child[i],keys,stk);
            stk.pop_back();
        }
    }
}

int main()
{
    string s;
    
    vector<string> vec;
    map<char,bool> m;
    bool o = false;
    while(true)
    {
        cin>>s;
        if (s[s.length()-1]=='#') 
        { 
            o = true;
            s.erase(s.length()-1);
        }
        vec.push_back(s);
        for (int j=0;j<s.length();++j)
            m[s[j]] = true;

        if (o) break;
    }

    vector<char> keys;
    map<char,bool>::iterator it;
    for (it = m.begin();it!=m.end();++it)
        keys.push_back(it->first);

    trienode* T = new trienode(keys.size());

    for (int i=0;i<vec.size();++i)
        insert(T,keys,vec[i]);
    
    vector<char>stk;
    print(T,keys,stk);

    return 0;
}
